node/pkg/driver/device_connectivity/device_connectivity_iscsi.go

```diff
@@ -32,6 +43,113 @@ func NewOsDeviceConnectivityIscsi(executer executer.ExecuterInterface) OsDeviceC
 	}
 }
 
+func (r OsDeviceConnectivityIscsi) iscsiCmd(args ...string) (string, error) {
+	out, err := r.Executer.ExecuteWithTimeout(int(iscsiCmdTimeout.Seconds()), "iscsiadm", args)
+	return string(out), err
+}
+
+func (r OsDeviceConnectivityIscsi) iscsiDiscover(portal string) error {
+	output, err := r.iscsiCmd("-m", "discoverydb", "-t", "sendtargets", "-p", portal, "--discover")
+	if err != nil {
+		logger.Errorf("Failed to discover iSCSI: {%s}, error: {%s}", output, err)
+		return err
+	}
+	return nil
+}
+
+func (r OsDeviceConnectivityIscsi) iscsiLogin(targetName, portal string) {
+	portalWithPort := portal + ":" + string(iscsiPort)
+	output, err := r.iscsiCmd("-m", "node", "-p", portalWithPort, "-T", targetName, "--login")
+	if err != nil {
+		logger.Errorf("Failed to login iSCSI: {%s}, error: {%s}", output, err)
+	}
+}
+
+func (r OsDeviceConnectivityIscsi) iscsiGetRawSessions() ([]string, error) {
+	output, err := r.iscsiCmd("-m", "session")
+	if err != nil {
+		if exitError, isExitError := err.(*exec.ExitError); isExitError && exitError.ExitCode() == iSCSIErrNoObjsFound {
+			logger.Debug("No active iSCSI sessions.")
+			return []string{}, nil
+		}
+		logger.Error("Failed to check iSCSI sessions.")
+		return nil, err
+	}
+	lines := strings.Split(strings.TrimSpace(output), "\n")
+	return lines, nil
+}
+
+func (r OsDeviceConnectivityIscsi) getAllSessions() (map[string]map[string]bool, error) {
+	lines, err := r.iscsiGetRawSessions()
+	if err != nil {
+		return nil, err
+	}
+	parseErr := errors.New("failed to parse iSCSI sessions")
+	portalsByTarget := make(map[string]map[string]bool)
+	for _, line := range lines {
+		parts := strings.Fields(line)
+		if len(parts) < 4 {
+			return nil, parseErr
+		}
+		portalInfo, targetName := parts[2], parts[3]
+		ipPortSeparatorIndex := strings.LastIndex(portalInfo, ":")
+		if ipPortSeparatorIndex < 0 {
+			return nil, parseErr
+		}
+		ip := portalInfo[:ipPortSeparatorIndex]
+		if set := portalsByTarget[targetName]; set == nil {
+			portalsByTarget[targetName] = make(map[string]bool)
+		}
+		portalsByTarget[targetName][ip] = true
+	}
+	return portalsByTarget, nil
+}
+
+func (r OsDeviceConnectivityIscsi) filterLoggedIn(portalsByTarget map[string][]string) (map[string][]string, error) {
+	loggedInPortalsByTarget, err := r.getAllSessions()
+	if err != nil {
+		return nil, err
+	}
+	filteredPortalsByTarget := make(map[string][]string)
+	for targetName, portals := range portalsByTarget {
+		for _, portal := range portals {
+			if !loggedInPortalsByTarget[targetName][portal] {
+				portals := filteredPortalsByTarget[targetName]
+				filteredPortalsByTarget[targetName] = append(portals, portal)
+			}
+		}
+	}
+	return filteredPortalsByTarget, nil
+}
+
+func (r OsDeviceConnectivityIscsi) iscsiDiscoverAny(portals []string) bool {
```

Comment:
please explain or add comments for this function, it is confusing.
why error != nil and return true?

Reply:
according to @igorgonibm, one portal discovery (per target name) is sufficient.
so this function tries to successfully discover any (one) portal.
if this confusing, could you specify what should I write to make it clearer?

Reply:
you fixed the bug